### ISSUE TITLE
GitHub + Capistrano Deploy issue

### ISSUE THREAD

**Issue by ThomasBush** [WestMichiganRubyTraining/discussion#45]
When attempting to deploy using Capistrano (for the first time) I get the following error.  `Permission denied (publickey).`  I assume this was a issue with Git/Github on the new server.  So I ssh'ed in using the Github guide link below.  After completing all steps in this guide I still get the error.  Now I am confused, is this a Github or Capistrano error?  If I ssh to my box I can run a `git pull` on this repo, which makes me think Capistrano, is where my mistake lies.  Any advice would be appreciated.     

help.github.com/articles/error-permission-denied-publickey

**Comment by jasonswett** [comment]
Have you looked at either of these two SO questions/answers? I don't know if either is the same problem as you, but looks like other people have had the same problem.

http://stackoverflow.com/questions/17457323/permission-denied-public-key-with-cap-but-i-can-ssh-into-github
http://stackoverflow.com/questions/7968656/why-is-a-cap-deploy-giving-permission-denied-publickey

## 

Jason Swett
IT Consultant
http://www.benfranklinlabs.com/
(616) 856-8075

See my article and video interview with the Wall Street Journal

**Comment by ThomasBush** [comment]
Jason, 

Thanks for the response.  I hadn't seen those questions, unfortunately neither solves my issue.  

The first turns on forward agent, which from my understanding meant that the app could only be deployed from my computer.  I work with another developer, and while I am currently the only one working on this, its still not ideal. Regardless of this, I tried turning this setting on as I figured if this worked that may be a clue, unfortunately it doesn't work.  

The second suggestion is to try that command `ssh -vT git@github.com` works fine when I ssh into the box, which leads me to believe it is a Capistrano setting I am missing as opposed to a GitHub error on my part.  I will focus my search on this assumption.

**Comment by billgathen** [comment]
How about this one? I went to "help" on GH, then pasted in the error.

https://help.github.com/articles/error-permission-denied-publickey

**Comment by ThomasBush** [comment]
Thanks @billgathen unfortunately I think that's the same GitHub article from the bottom of my initial post, and I can complete all the steps successfully in that walkthrough. Which has led me to believe my mistake was within a deploy setting in Capistrano.

**Comment by billgathen** [comment]
Sorry, I'm an idiot. I overlooked it because it wasn't a link.

Can you drop your capistrano file in here, or into a gist? Redact any proprietary info first. :wink:

**Comment by ThomasBush** [comment]
@billgathen , no worries, must have forgotten to link, sorry.  Here is my gist https://gist.github.com/ThomasBush/8f214ab34a0ced2fe3bf

**Comment by billgathen** [comment]
You said you could connect properly to github from your machine. Have you confirmed that you can get there from your destination server? Capistrano ssh's into your destination server and runs the commands from there. If there's no ssh key that GH recognizes on the destination server, you'd get that error.

**Comment by billgathen** [comment]
Alternately, you could use `set :deploy_via :copy` and capistrano will pull down the repo to your local machine, zip it, push the zip up to the destination box and unzip it. Then the destination box doesn't need the key.

We had to take that approach when we moved into AWS and the servers up there couldn't see our private git repo.

Putting the key out there would be a better solution, though, as it's simpler and more efficient.

**Comment by thejbsmith** [comment]
@ThomasBush, can you post a gist of the entire output when you run a cap deploy with the --trace flag?

For the server to connect to your Github repository, did you set up the server with it's own SSH key? (I would assume so since you aren't using agent forwarding). If so, is the SSH key set up for the user you are trying to deploy as?

AWS servers won't inherently have any issues with connecting to a private repo (I'm currently deploying to multiple AWS servers from a private Git repo). But this can definitely change depending on how the machine is set up, how the firewalls are handled, which ports are open, etc.

Also, if possible I would try to stay away from using `set :deploy_via :copy` as it is a much slower method of deploying (but does have it's place in certain situations). But, as @billgathen pointed out, using `set :deploy_via :copy` should be a quick workaround for your current issue as your server would not have to talk to Github.

Another option I have used in the past is Git Deploy Keys (https://help.github.com/articles/managing-deploy-keys#deploy-keys). Basically you can set up a repo with a specific SSH key to deploy with. The link lists some of the pros and cons, as well as other ways to manage keys.

Just a note on SSH agent forwarding:
Turning on SSH agent forwarding should work for any machine that has the proper SSH keys in place, not just your machine (e.g. you have another developer that can SSH into the server, this should work for him as well).

**Comment by ThomasBush** [comment]
@billgathen attempting to set `:deploy_via, :copy` renders the same error, which doesn't really make sense, right?

@thejbsmith, I have included the output of `cap production deploy --trace` in the gist below.
https://gist.github.com/ThomasBush/f8036033b9d5fd34cef3

To answer your questions, yes I set up ssh keys on the remote server.  This key was generated for my deploy user.  I ssh'ed onto the server and ran `ssh-keygen -lf ~/.ssh/id_rsa.pub` which matches up with the key in GitHub.  I have also ssh'ed into this server and run the `git clone` and `git pull` commands, would that mess this up having the repo there already?

I used the same link you provided - Git Deploy Keys (https://help.github.com/articles/managing-deploy-keys#deploy-keys) to arrive at my current setup, only I went with the Machine Users method outlined directly below the section you have highlighted.  

Not sure if this is relevant, but in an effort to provide as much info as possible: I created an organization "HexArmor" containing the private repo "HexGen" and a "Deployment" Team.  I also created a machine user "hexarmor-deploy" who is a member of the deployment team which has read access to the private repo "HexGen" and ssh keys which match the servers ssh keys as mentioned above.  

I noticed mention of AWS so for a bit more clarity I am also using a RackSpace cloud server, running Ubuntu 12.04 LTS, which I spun up with rvm and Phusion Passenger.  
- Ruby 1.9.3p484
- Rails 3.2.13
- rvm 1.24.7
- Capistrano 3.0.1

Thanks for all the help, I really appreciate it, also please let me know if I missed a question in there or there is more information I can provide.

**Comment by thejbsmith** [comment]
It looks like the command that it is failing on is `GIT_ASKPASS=/bin/echo GIT_SSH=/tmp/HexGen/git-ssh.sh /usr/bin/env git ls-remote git@github.com:HexArmor/hexgen.git`. Can you SSH into the server and try to run that exact command? If that succeeds, can you try to run the same command with `sudo` and see if it still succeeds?

Capistrano usually runs everything via sudo, so depending on the setup this can cause issues. If the command above works when running as the user specified in the deploy script but not when running sudo, try adding `set :use_sudo, false` to your deploy file and try deploying again.

If that doesn't help, I would be happy to sit down with you and try to get this figured out. Sitting in front of a computer debugging this in real time is probably going to be the best solution at this point.

**Comment by ThomasBush** [comment]
@thejbsmith 

I ssh'ed into my server and ran the command:

`GIT_ASKPASS=/bin/echo GIT_SSH=/tmp/HexGen/git-ssh.sh /usr/bin/env git ls-remote git@github.com:HexArmor/hexgen.git`

With and without `sudo` it returns the same output which I have included in the following gist
https://gist.github.com/ThomasBush/481c60783f98dd62ba61

If you wouldn't mind meeting I would really appreciate it, I'd be happy to buy you a meal, a drink, or whatever.  Just let me know.  Thanks!

**Comment by thejbsmith** [comment]
@ThomasBush, send me an email at jim@jimsmithdesign.com and we'll figure out something. After we figure out the issue we'll post here with the solution.

**Comment by ThomasBush** [comment]
Thanks, @thejbsmith for helping me work through this last night.

So to summarize this in hopes it will help someone else, there were a few issues that we needed to correct.  
- I originally generated the ssh key on the server with a password.  Removing this, and regenerating without a password, than linking this in GitHub was the solution here.
- I had my `set :user, "enter-username"` set as my GitHub deploy user.  This should be the ssh user used the access the server, not the GitHub user.  So switching to the correct user solved this issue.
- MySQL was installed on the server, but some how I forgot to actually start it?!  I was happy @thejbsmith caught this one quickly because the error read like a user or password was wrong.  We just ran `service mysql start` and created the db with the proper name to solve this issue.
- Also the repo was on the server, I did this to test git, github, and credentials were all correctly set up.  Capistrano needed to work from a clean slate though.  So removing this was the solution here and changing my vhost files to point to the new correct folder.  Capistrano versions by placing a "current" folder, so simply adding this path addition to my /etc/apache2/sites-available/{my-named-vhost} file solved this issue. 

Those were the major issues, hopefully that helps someone.  Again big thanks to @thejbsmith as I think this took 2-3 hours to work through these and a few other minor issues.  I was talking to people at work yesterday and today and we all agreed it was pretty cool to be a part of a community that is willing to help to that extent.  Thanks @billgathen and @jasonswett for helping work through this as well.
